storage/innobase/xtrabackup/src/xbcloud/xbcloud.cc

```diff
@@ -836,29 +888,91 @@ bool xbcloud_put(Object_store *store, const std::string &container,
       filehash.erase(entry->path);
     }
 
+    /* Reset chunk */
     memset(&chunk, 0, sizeof(chunk));
-  } while (!has_errors);
+  } while (!cntx->has_errors->load());
 
   h.stop();
-  thread.join();
+  ev.join();
 
-  xb_stream_read_done(stream);
+end:
+  if (stream != nullptr) xb_stream_read_done(stream);
+  pthread_exit(NULL);
+}
 
+bool xbcloud_put(Object_store *store, const std::string &container,
+                 const std::string &backup_name) {
+  bool exists;
+  put_thread_ctxt_t *data_threads;
+  std::atomic<bool> has_errors{false};
+  pthread_t *threads;
+  if (!store->container_exists(container, exists)) {
+    return false;
+  }
 
-  // check if backup directory exists and it has some files
-  if (res == XB_STREAM_READ_ERROR || has_errors ||
+  if (!exists) {
+    if (!store->create_container(container)) {
+      return false;
+    }
+  }
+
+  std::vector<std::string> object_list;
+  if (!store->list_objects_in_directory(container, backup_name, object_list)) {
+    return false;
+  }
+
+  if (!object_list.empty()) {
+    msg_ts("%s: error: backup named %s already exists!\n", my_progname,
+           backup_name.c_str());
+    return false;
+  }
+
+  /* Create data copying threads */
+  data_threads = (put_thread_ctxt_t *)my_malloc(
+      PSI_NOT_INSTRUMENTED, sizeof(put_thread_ctxt_t) * (opt_threads + 1),
+      MYF(MY_FAE));
+  threads = new pthread_t[opt_threads];
+  for (uint i = 0; i < (uint)opt_threads; i++) {
+    data_threads[i].num = i;
+    data_threads[i].has_errors = &has_errors;
+    data_threads[i].store = store;
+    data_threads[i].container = &container;
+    data_threads[i].buf_md5 = new Http_buffer();
+    pthread_create(&threads[i], nullptr, &put_func, data_threads + i);
```

Comment:
handle the error if pthread_create fails.

Reply:
Done. `has_errors` will signal already started threads to cleanup and exit.